phlib/nativethread.c

```diff
@@ -811,32 +811,204 @@ NTSTATUS PhGetThreadLastStatusValue(
 }
 
 /**
- * Retrieves the COM apartment of a thread.
+ * Retrieves statistics about multi-threaded apartment usage in a process.
  *
- * \param ThreadHandle A handle to the thread.
- * \param ProcessHandle A handle to the process.
- * \param ApartmentState The COM apartment of the thread.
+ * \param[in] ProcessHandle A handle to the process. The handle must have
+ * PROCESS_QUERY_LIMITED_INFORMATION and PROCESS_VM_READ access.
+ * \param[out] MTAInits The total number of MTA references in the process.
+ * \param[out] MTAIncInits The number of MTA references from CoIncrementMTAUsage.
  *
  * \return Successful or errant status.
  */
-NTSTATUS PhGetThreadApartmentState(
+NTSTATUS PhGetProcessMTAUsage(
+    _In_ HANDLE ProcessHandle,
+    _Out_opt_ PULONG MTAInits,
+    _Out_opt_ PULONG MTAIncInits
+    )
+{
+    NTSTATUS status;
+#ifdef _WIN64
+    BOOLEAN isWow64;
+#endif
+    static PH_INITONCE initOnce = PH_INITONCE_INIT;
+    static PMTA_USAGE_GLOBALS mtaUsageGlobals = NULL;
+
+    if (!MTAInits && !MTAIncInits)
+        return STATUS_INVALID_PARAMETER;
+
+#ifdef _WIN64
+    if (!NT_SUCCESS(status = PhGetProcessIsWow64(ProcessHandle, &isWow64)))
+        return status;
+
+    if (isWow64)
+        return STATUS_NOT_SUPPORTED;
+#endif
+
+    if (PhBeginInitOnce(&initOnce))
+    {
+        PVOID combase;
+        PMTA_USAGE_GLOBALS (WINAPI* CoGetMTAUsageInfo_I)(VOID);
+
+        if (combase = PhGetLoaderEntryDllBaseZ(L"combase.dll"))
+        {
+            // combase exports CoGetMTAUsageInfo as ordinal 70 since Windows 8
```

Comment:
Should this be guarded by `>= WINDOWS_8`?

Reply:
`combase.dll` itself exists since Windows 8, so `CoGetMTAUsageInfo` is always available there. But yeah, I'll add a version check to avoid unsuccessful loading attempts on earlier versions.

Reply:
Thanks, I was mostly concerned if the ordinal was in use prior to Windows 8. I've seen ordinal change and get reused, it's rare, but happens.